app/src/lib/stores/app-store.ts

```diff
@@ -3847,18 +3857,72 @@ export class AppStore extends TypedBaseStore<IAppState> {
     this.hasUserViewedStash = false
   }
 
-  private async refreshAfterCheckout(repository: Repository, branch: Branch) {
+  /**
+   * @param commitish A branch name or a commit hash
+   */
+  private async refreshAfterCheckout(
+    repository: Repository,
+    commitish: string
+  ) {
     this.updateCheckoutProgress(repository, {
       kind: 'checkout',
       title: `Refreshing ${__DARWIN__ ? 'Repository' : 'repository'}`,
       value: 1,
-      targetBranch: branch.name,
+      target: commitish,
     })
 
     await this._refreshRepository(repository)
     return repository
   }
 
+  /**
+   * Checkout the given commit, ignoring any local changes.
+   *
+   * Note: This shouldn't be called directly. See `Dispatcher`.
+   */
+  public async _checkoutCommit(
+    repository: Repository,
+    commit: CommitOneLine
+  ): Promise<Repository> {
+    const repositoryState = this.repositoryStateCache.get(repository)
+    const { branchesState } = repositoryState
+    const { tip } = branchesState
+
+    // No point in checking out the currently checked out commit.
+    if (tip.kind === TipState.Valid && tip.branch.tip.sha === commit.sha) {
+      return repository
+    }
+
+    return this.withAuthenticatingUser(repository, (repository, account) => {
+      // We always want to end with refreshing the repository regardless of
+      // whether the checkout succeeded or not in order to present the most
+      // up-to-date information to the user.
+      return this.checkoutCommitDefaultBehaviour(repository, commit, account)
+        .catch(e =>
+          this.emitError(
+            // new CheckoutError(e, repository, branch)
+
+            new ErrorWithMetadata(e, {
+              gitContext: { kind: 'checkout', commitToCheckout: commit },
+              repository,
+            })
+          )
+        )
```

Comment:
```suggestion
        .catch(e => this.emitError(new Error(e)))
```

In the `_checkoutBranch` method, the `CheckoutError` extending the `ErrorWithMetadata` is currently only needed to pass some metadata for specific error handling implemented for checking out a branch that is not handled in this PR (such as providing the ability to stash before checking out).